x-pack/auditbeat/internal/ebpf/watcher_linux.go

```diff
@@ -0,0 +1,154 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build linux
+
+package ebpf
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/elastic/ebpfevents"
+)
+
+var (
+	gWatcherOnce sync.Once
+	gWatcherErr  error
+	gWatcher     watcher
+)
+
+type client struct {
+	name    string
+	mask    EventMask
+	records chan ebpfevents.Record
+}
+
+type watcher struct {
+	sync.Mutex
+	ctx     context.Context
+	cancel  context.CancelFunc
+	loader  *ebpfevents.Loader
+	clients map[string]client
+	status  status
+}
+
+type status int
+
+const (
+	stopped status = iota
+	started
+)
+
+func GetWatcher() (Watcher, error) {
+	gWatcher.Lock()
+	defer gWatcher.Unlock()
+
+	// Try to load the probe once on startup so consumers can error out.
+	gWatcherOnce.Do(func() {
+		if gWatcher.status == stopped {
+			l, err := ebpfevents.NewLoader()
+			if err != nil {
+				gWatcherErr = fmt.Errorf("init ebpf loader: %w", err)
+				return
+			}
+			_ = l.Close()
+		}
+	})
+
+	return &gWatcher, gWatcherErr
+}
+
+func (w *watcher) Subscribe(name string, events EventMask) <-chan ebpfevents.Record {
+	w.Lock()
+	defer w.Unlock()
+
+	if w.status == stopped {
+		startLocked()
+	}
+
+	w.clients[name] = client{
+		name:    name,
+		mask:    events,
+		records: make(chan ebpfevents.Record),
```

Comment:
This kills the buffering from the two channels before, and I'm not sure it was intentional, so please correct me if I'm wrong:
- flow is: ebpfevents -> watcher channel -> fan_out(client channels)
- ebpfevents exposes its internal buffer length which you use in the watcher channel https://github.com/elastic/beats/blob/a5986dd20bbae0d57e672d996d731290aa6301e3/x-pack/auditbeat/internal/ebpf/watcher_linux.go#L102
- Then the watcher channel fans it to all clients https://github.com/elastic/beats/blob/a5986dd20bbae0d57e672d996d731290aa6301e3/x-pack/auditbeat/internal/ebpf/watcher_linux.go#L111
- The client channels are synchronous, so you have `buffered(ebpfevents) -> buffered(watcher) -> synchronous(client)` . Effectively you only drain the watcher (buffered) channel once you synchronize with all clients once, essentially removing the previous buffering. In other words: if one client too slow to read its channel, you stop draining the watcher, which pushes the back-pressure to ebpfevents which now will block on the watcher, which "can" in turn start dropping events from kernel->ebpfevents.

If this was not intentional, I'd suggesting using the same bufferlen on each client channel as well.

Reply:
Thanks for the good explaination of this. It wasn't intentional, and I've changed the client channels to be buffered